dsl/pace/dsl/dace/orchestration.py

```diff
@@ -36,16 +36,23 @@
 from pace.util.mpi import MPI
 
 
+try:
+    import cupy as cp
+except ImportError:
+    cp = None
+
+
 def dace_inhibitor(func: Callable):
     """Triggers callback generation wrapping `func` while doing DaCe parsing."""
     return func
 
 
 def _upload_to_device(host_data: List[Any]):
     """Make sure any data that are still a gt4py.storage gets uploaded to device"""
-    for data in host_data:
-        if isinstance(data, gt4py.storage.Storage):
-            data.host_to_device()
+    assert cp is not None
+    for i, data in enumerate(host_data):
+        if isinstance(data, cp.ndarray):
+            host_data[i] = cp.asarray(data)
```

Comment:
This might need to work in both cases: whether it's a CPU or GPU backend.

Reply:
Yeah, I think this changes the code in a way that will break it. I do wonder though whether this function is needed at all. It's called only when parsing/calling dace programs, to convert input arguments to gpu data to pass to dace. However, all storage data should already be allocated on gpu in that case, so this call should be redundant. This is evidenced by it doing nothing to Quantity, and all arguments to the dace programs should be Quantity, not Storage.

Reply:
I think I fixed this by looking at all references and checking for `config.is_gpu_backend()` before calling. I'll also add a note here that this will raise an exception if it is called otherwise.